### PR TITLE
Update --hires_fix

### PR DESCRIPTION
**Pull request by JPPhoto**
Change `--hires_fix` to calculate initial width and height based on the model's resolution (if available, else 512x512) and with a minimum dimension equal to 0.5 * resolution. Because diffusers isn't bound to resolutions in increments of 64, this resizing isn't, either.

### CODE REVIEW

File: ldm/invoke/generator/txt2img2img.py [hunk 1]
```diff
@@ -38,10 +41,32 @@ def get_make_image(self, prompt:str, sampler, steps:int, cfg_scale:float, ddim_e
...
+        if aspect > 1.0:
+            init_height = max(min_dimension, math.sqrt(model_area / aspect))
+            init_width = init_height * aspect
+        else:
+            init_width = max(min_dimension, math.sqrt(model_area * aspect))
+            init_height = init_width / aspect
+
+        init_width, init_height = trim_to_multiple_of(math.floor(init_width), math.floor(init_height))
+        self.init_width = init_width
+        self.init_height = init_height
```
Comment (keturn) [inline ldm/invoke/generator/txt2img2img.py:68]: I think it's better in this case to _not_ store these as instance attributes, and let the nested `make_image` function refer to the values from its outer scope instead.

Reason being that I think the instance outlives any one `get_make_image`/`get_image` call, and you don't want the different `get_image` functions to refer to that state that might be changing.

(The whole nested `make_image` function confuses things a lot, and I think it's the next thing on the chopping block after the 2.3 release.)

**Review by keturn — DISMISSED** [review]
Hmm. I don't have a thorough enough understanding of the model to say whether keeping the _area_ constant is more important than keeping a single linear _axis_ constant.

I'd be a little surprised if `dimension * 0.5` works well, but I really haven't tested that assumption much. Maybe we can put together a prompt file to try some things out at different aspect ratios. We wouldn't even need to run them through the whole txt2img2img, just the first pass would be enough to see if it's repeating elements.

File: ldm/invoke/generator/txt2img2img.py (hunk 1 above)
Comment (JPPhoto) [reply ldm/invoke/generator/txt2img2img.py:68]: Relevant lines:
https://github.com/invoke-ai/InvokeAI/blob/d57406c4fe6a73f264f8c3663a427ddf3cb07a81/ldm/invoke/generator/base.py#L72-L87

Comment (JPPhoto) [reply ldm/invoke/generator/txt2img2img.py:68]: Regarding duplication, my testing has shown that any dimension larger than 512 will yield duplication, but those that are too small will cause artifacts in the output. That's how I arrived at 0.5 as a balance and the area calculation as a compromise. Truly if you want to generate an ultra-wide or ultra-tall image and maintain control over the output, the canvas is the only way to go.

Sample outputs, same prompt and seed, fixed maximum dimension of 512:

512x64
512x128
512x192
512x256
512x320
512x384
512x448
512x512
448x512
384x512
320x512
256x512
192x512
128x512
64x512

**Review by lstein — APPROVED** [review]
I tested a series of extremely skewed aspect ratios, e.g. 512x1152 and 1152x512. They generate reproducible image doublings without `--hires_fix` and generate non-doubled images with the option. So it appears to work.

However, when one dimension is less than 512, then `--hires_fix` does not work so well, and there is still doubling behavior. I think the issue is that the small image generated automatically has doubling under this scenario, and this can't be fixed easily. This issue won't come up frequently.

**Comment by JPPhoto** [comment]
@lstein One option for generating smaller images is to use the `--hires_fix` logic and then downsample instead of doing the second generation pass. That's worth considering for another PR.

**Comment by lstein** [comment]
@keturn I think that @JPPhoto has made a commit that addresses your change request. If this is acceptable to you, could you change your review to approve?

File: ldm/invoke/generator/txt2img2img.py [hunk 2]
```diff
@@ -38,10 +41,32 @@ def get_make_image(self, prompt:str, sampler, steps:int, cfg_scale:float, ddim_e
...
+        # Scale the input width and height for the initial generation
+        # Make their area equivalent to the model's resolution area (e.g. 512*512 = 262144),
+        # while keeping the minimum dimension at least 0.5 * resolution (e.g. 512*0.5 = 256)
+
+        aspect = width / height
+        dimension = 512
+        try:
+            dimension = self.model.unet.config.sample_size * self.model.vae_scale_factor
+        except:
+            pass
```
Comment (keturn) [inline ldm/invoke/generator/txt2img2img.py:54]: Why is there a try/except around this arithmetic?

**Comment by keturn** [comment]
On a closer reading, the `init_width`, `init_height`, and `scale` values calculated in `get_make_image` were not used anywhere, except for this debug message: https://github.com/invoke-ai/InvokeAI/blob/c18db4e47b10cf1658612f3eec2d537a789b10ea/ldm/invoke/generator/txt2img2img.py#L58

which is nothing but lies through and through, because those values don't come from anything attached to what it sent to or received from the pipeline, nor does it set the scheduler to DDIM anywhere.

In reality, it uses the size of the `x_T` passed to `make_image`, and as you've pointed out, that is the result of the `get_noise` method.

Question: Are the `width` and `height` values passed to `get_noise` any less reliable than the ones passed to `get_make_image`?

Because if they're just as good, then we can leave this calculation to be _only_ in the `if scale` block of `get_noise`, and avoid making up a side channel to communicate values from one method to the other.

**Comment by JPPhoto** [comment]
> Question: Are the `width` and `height` values passed to `get_noise` any less reliable than the ones passed to `get_make_image`?
> 
> Because if they're just as good, then we can leave this calculation to be _only_ in the `if scale` block of `get_noise`, and avoid making up a side channel to communicate values from one method to the other.

Looking at the code, there may be side effects from moving the width and height calculation there. `width` and `height` are passed around and currently represent the *final* width and height of the generation. I think it's worth refactoring/investigating after this gets merged, but I can look more into it prior if you think that's the way to go.

**Comment by JPPhoto** [comment]
@keturn I did some testing and moving the calculation to the noise generation function was fine. I checked that in here.

**Pull request merged**